subprojects/core/src/main/java/org/gradle/configuration/DefaultScriptPluginFactory.java

```diff
@@ -124,7 +130,7 @@ public void apply(final Object target) {
                     return fileLookup.getFileResolver().getPatternSetFactory();
                 }
             };
-            services.add(ScriptPluginFactory.class, DefaultScriptPluginFactory.this);
+            services.add(ScriptPluginFactory.class, scriptPluginFactory);
```

Comment:
One way to detangle this would be to pass this service registry from the outside instead of building it here. Would also reduce the number of constructor arguments.